docs/api: Improve Golang API documentation (#3948)
inspektor-gadget/inspektor-gadget: docs/api/golang.mdx

    "mntns_id": 4026534339,
    "parent": {
      "comm": "sh",
      "pid": 42678
    },
    "pid": 53660,
    "tid": 53660
  },
  "runtime": {
    "containerId": "1b0ba0f6f7f7251a8dedbee0a836dc3edadeba7444cd56d6afe0989b6007bcd7",
    "containerImageDigest": "sha256:2919d0172f7524b2d8df9e50066a682669e6d170ac0f6a49676d54358fe970b5",
    "containerImageName": "busybox",
    "containerName": "mycontainer",
    "containerPid": 42678,
    "containerStartedAt": 1737743296076978537,
    "runtimeName": "docker"
  },
  "timestamp": "2025-01-24T13:37:35.312976962-05:00",
  "timestamp_raw": 1737743855312976962
}
```

As expected, only events from `mycontainer` are captured and the runtime leaf
contains all information related to the container creating the events. Also, a
new timestamp field with a human friendly version is present.

Don't forget to remove the containers create above:

```bash
$ docker rm mycontainer foocontainer
```

### Data Source

So far, we've been using the json formatter to convert the events produced by a
Gadget to json. However, if you're only interested on getting specific fields,
it can be more performant to access them directly using a Field Accessor instead

> of performing a full json marshaling and unmarshaling.

Author: We don't. Marshaling to json requires converting accessing all fields in the data, so I suspect it's going to be always slower, but perhaps @flyth can provide some insights here.

> of performing a full json marshaling and unmarshaling.

Reviewer: Yeah, exactly - accessors are usually directly accessing the memory of a specific (and only that) field.

command systemd-oomd (1245) opened /proc/meminfo
command systemd-oomd (1245) opened /sys/fs/cgroup/user.slice/user-1001.slice/user@1001.service/memory.pressure
command systemd-oomd (1245) opened /sys/fs/cgroup/user.slice/user-1001.slice/user@1001.service/memory.current
command systemd-oomd (1245) opened /sys/fs/cgroup/user.slice/user-1001.slice/user@1001.service/memory.min
command systemd-oomd (1245) opened /sys/fs/cgroup/user.slice/user-1001.slice/user@1001.service/memory.low
command systemd-oomd (1245) opened /sys/fs/cgroup/user.slice/user-1001.slice/user@1001.service/memory.swap.current
command systemd-oomd (1245) opened /sys/fs/cgroup/user.slice/user-1001.slice/user@1001.service/memory.stat
command systemd-oomd (1245) opened /proc/meminfo
command dbus-daemon (3811) opened /sys/kernel/security/apparmor/.access
command tmux: server (12404) opened /proc/38122/cmdline
command systemd-oomd (1245) opened /proc/meminfo
command systemd-oomd (1245) opened /proc/meminfo
command tmux: server (12404) opened /proc/38122/cmdline
```

### Running Gadgets remotely with the gRPC runtime

All the examples above run the eBPF locally. This example shows how to control a

> remote `ig` instance by using the gRPC runtime.

Reviewer 2: You could add some details:
- The client does not need connectivity to the OCI registry.
- The client does not need credentials to the OCI registry.

Is it correct btw?

  Use to run a Gadget in the local host. It's only supported in Linux.
- [grpc](https://pkg.go.dev/github.com/inspektor-gadget/inspektor-gadget@%IG_BRANCH%/pkg/runtime/grpc):
  Use to run a Gadget in a remote Linux host. It's supported in any OS.

### Operators

Operators are the cornerstones of the Inspektor Gadget architecture. They are
modular components responsible for specific tasks, such as loading eBPF
programs, enriching events, sorting output, etc. Inspektor Gadget provides
different operators as described in the [operators](../spec/operators/)
documentation. It is also possible to create custom operators as explained
below.

### Data Sources

A [Data
Source](https://pkg.go.dev/github.com/inspektor-gadget/inspektor-gadget@%IG_BRANCH%/pkg/datasource)
is a mechanism used to transport the data generated by the Gadgets across the
different components. Operators can consume and mutate the data flowing through
a data source.

### Gadget Context

The [Gadget
Context](https://pkg.go.dev/github.com/inspektor-gadget/inspektor-gadget@%IG_BRANCH%/pkg/gadget-context)
is the component that glues everything together. This carries information

> between operators, runtimes, etc.

Author: Yeah, it could help. Unfortunately I'm not completely sure about how to make it.

$ sudo ig image export trace_open:latest trace_open.tar
```

import FromFile from '!!raw-loader!./_gadgets/from_file/main.go';

<CodeBlock language="go">{FromFile}</CodeBlock>

This example is run and produces the same results as above.

### Running a Gadget embedded in your application

The example above requires you to ship the tarball file with your application
binary, it's not flexible, so let's explore how to embeed the Gadget image on
your binary by using `//go:embed`.

import FromMemory from '!!raw-loader!./_gadgets/from_memory/main.go';

<CodeBlock language="go">{FromMemory}</CodeBlock>

Again, this is run and will produce the same result as above.

### Using operators from Inspektor Gadget

In this example we use some operators provided by Inspektor Gadget to extend the
functionality of your application:

- [LocalManager](../spec/operators/localmanager.md): Enriches events with
  container information. Also provides a mechanism to filter events by
  container.
- [Formatters](../spec/operators/formatters.md): Provides a pretty
  representation of some events like timestamps, etc.

Parameters for each operator are passed as a map of strings using the

> fully qualified name of the parameter as the key.

Reviewer: We should link to and explain what a "fully qualified name" is sooner or later.

> import Operators from '!!raw-loader!./_gadgets/operators/main.go';

Author: I see the file at  `docs/api/_gadgets/operators/main.go`  and it's rendering fine:

![image](https://github.com/user-attachments/assets/b91a51ac-9322-45e7-a8af-69c52f2f68fe)

```bash
$ go run -exec sudo .
{
  "error": "",
  "error_raw": 0,
  "fd": 3,
  "flags_raw": 0,
  "fname": "/dev/null",
  "k8s": {
    "containerName": "",
    "hostnetwork": false,
    "namespace": "",
    "node": "",
    "owner": {
      "kind": "",
      "name": ""
    },
    "podLabels": "",
    "podName": ""
  },
  "mode_raw": 0,
  "proc": {
    "comm": "cat",
    "creds": {
      "gid": 0,
      "uid": 0
    },
    "mntns_id": 4026534339,
    "parent": {
      "comm": "sh",
      "pid": 42678
    },
    "pid": 53660,
    "tid": 53660
  },
  "runtime": {
    "containerId": "1b0ba0f6f7f7251a8dedbee0a836dc3edadeba7444cd56d6afe0989b6007bcd7",
    "containerImageDigest": "sha256:2919d0172f7524b2d8df9e50066a682669e6d170ac0f6a49676d54358fe970b5",
    "containerImageName": "busybox",
    "containerName": "mycontainer",
    "containerPid": 42678,
    "containerStartedAt": 1737743296076978537,
    "runtimeName": "docker"
  },
  "timestamp": "2025-01-24T13:37:35.312976962-05:00",
  "timestamp_raw": 1737743855312976962
}
```

As expected, only events from `mycontainer` are captured and the runtime node
contains all information related to the container creating the events. Also, a

> new timestamp field with a human friendly version is present.

Reviewer: ```suggestion
new timestamp field with a human-readable version is present.
```

Don't forget to remove the containers created above:

```bash
$ docker rm mycontainer foocontainer
```

### Data Source

So far, we've been using the json formatter to convert the events produced by a

> Gadget to json. However, if you're only interested on getting specific fields,

Reviewer: ```suggestion
Gadget to json. However, if you're only interested in getting specific fields,
```

command tmux: server (12404) opened /proc/38122/cmdline
command systemd-oomd (1245) opened /proc/meminfo
command systemd-oomd (1245) opened /sys/fs/cgroup/user.slice/user-1001.slice/user@1001.service/memory.pressure
command systemd-oomd (1245) opened /sys/fs/cgroup/user.slice/user-1001.slice/user@1001.service/memory.current
command systemd-oomd (1245) opened /sys/fs/cgroup/user.slice/user-1001.slice/user@1001.service/memory.min
command systemd-oomd (1245) opened /sys/fs/cgroup/user.slice/user-1001.slice/user@1001.service/memory.low
command systemd-oomd (1245) opened /sys/fs/cgroup/user.slice/user-1001.slice/user@1001.service/memory.swap.current
command systemd-oomd (1245) opened /sys/fs/cgroup/user.slice/user-1001.slice/user@1001.service/memory.stat
command systemd-oomd (1245) opened /proc/meminfo
command dbus-daemon (3811) opened /sys/kernel/security/apparmor/.access
command tmux: server (12404) opened /proc/38122/cmdline
command systemd-oomd (1245) opened /proc/meminfo
command systemd-oomd (1245) opened /proc/meminfo
command tmux: server (12404) opened /proc/38122/cmdline
```

### Running Gadgets remotely with the gRPC runtime

> All the examples above run the eBPF locally. This example shows how to control a

Reviewer: ```suggestion
All the examples above run the gadgets (and their eBPF parts) locally. This example shows how to control a
```

remote `ig` instance by using the gRPC runtime. In this case, the Gadget image
is pulled by the remote instance, hence the client doesn't need to contact the
OCI registry directly.

import GRPC from '!!raw-loader!./_gadgets/grpc/main.go';

<CodeBlock language="go">{GRPC}</CodeBlock>

> In this case we need to run `ig` on daemon mode:

Reviewer: ```suggestion
In this case we need to run `ig` in daemon mode:
```

```bash
$ sudo ig daemon --host tcp://127.0.0.1:8888
INFO[0000] starting Inspektor Gadget daemon at "tcp://127.0.0.1:8888"
```

> Now run the example. Note that in this case root is not needed because the all
> privileged operations are done by the `ig` process.

Reviewer: ```suggestion
Now run the example. Note that in this case running as user "root" is not needed, because all
the privileged operations are happening inside the `ig` process on the remote side.
```

```bash
$ go run .
{"comm":"irqbalance","err":0,"fd":6,"flags":0,"fname":"/proc/interrupts","gid":0,"mntns_id":4026533158,"mode":0,"pid":1262,"timestamp":6114008072805,"uid":0}
{"comm":"irqbalance","err":0,"fd":6,"flags":0,"fname":"/proc/stat","gid":0,"mntns_id":4026533158,"mode":0,"pid":1262,"timestamp":6114008961935,"uid":0}
{"comm":"irqbalance","err":0,"fd":6,"flags":0,"fname":"/proc/irq/65/smp_affinity","gid":0,"mntns_id":4026533158,"mode":0,"pid":1262,"timestamp":6114009113261,"uid":0}
{"comm":"irqbalance","err":0,"fd":6,"flags":0,"fname":"/proc/irq/98/smp_affinity","gid":0,"mntns_id":4026533158,"mode":0,"pid":1262,"timestamp":6114009138809,"uid":0}
```

### Other Examples

> There are other examples available on the repo, please check them at

Reviewer: ```suggestion
There are other examples available in the repo, please check them at
```
